rewrite-gradle/src/main/java/org/openrewrite/gradle/internal/GradleWrapperScriptDownloader.java

```diff
@@ -0,0 +1,114 @@
+
+public class GradleWrapperScriptDownloader {
+    private static final Path WRAPPER_SCRIPTS = Paths.get("rewrite-gradle/src/main/resources/META-INF/rewrite/gradle-wrapper/");
+
+    public static void main(String[] args) throws IOException, InterruptedException {
+        Lock lock = new ReentrantLock();
+        InMemoryExecutionContext ctx = new InMemoryExecutionContext();
+        List<GradleWrapper.GradleVersion> allGradleReleases = GradleWrapper.listAllVersions(null, ctx);
+        Map<String, GradleWrapperScriptLoader.Version> allDownloadedVersions =
+                new ConcurrentHashMap<>(new GradleWrapperScriptLoader().getAllVersions());
+
+        AtomicInteger i = new AtomicInteger();
+        ForkJoinPool pool = ForkJoinPool.commonPool();
+        pool.invokeAll(allGradleReleases.stream().map(version -> (Callable<Void>) () -> {
+            String v = version.getVersion();
+            String threadName = " [" + Thread.currentThread().getName() + "]";
+
+            if (allDownloadedVersions.containsKey(v)) {
+                System.out.printf("%03d: %s already exists. Skipping.%s%n", i.incrementAndGet(),
+                        v, threadName);
+                return null;
+            }
+
+            try {
+                DistributionInfos infos = DistributionInfos.fetch(GradleWrapper.DistributionType.Bin, version, ctx);
+                GradleWrapper wrapper = new GradleWrapper(v, infos);
+
+                String gradlewChecksum = downloadScript(WRAPPER_SCRIPT_LOCATION, wrapper, "unix", ctx);
+                String gradlewBatChecksum = downloadScript(WRAPPER_BATCH_LOCATION, wrapper, "windows", ctx);
+
+                lock.lock();
+                allDownloadedVersions.put(v, new GradleWrapperScriptLoader.Version(v, gradlewChecksum, gradlewBatChecksum));
+
+                List<String> sortedVersions = new ArrayList<>(allDownloadedVersions.keySet());
+                sortedVersions.sort(new LatestRelease(null).reversed());
+                try (BufferedWriter writer = Files.newBufferedWriter(WRAPPER_SCRIPTS.resolve("versions.csv"))) {
+                    writer.write("version,gradlew,gradlewBat\n");
+                    for (String sortedVersion : sortedVersions) {
+                        GradleWrapperScriptLoader.Version version1 = allDownloadedVersions.get(sortedVersion);
+                        writer.write(sortedVersion + "," + version1.getGradlewChecksum() + "," + version1.getGradlewBatChecksum() + "\n");
+                    }
+                }
+                System.out.printf("%03d: %s downloaded.%s%n", i.incrementAndGet(), v, threadName);
+            } catch (Throwable t) {
+                // FIXME There are some wrappers that are not downloading successfully. Why?
+                System.out.printf("%03d: %s failed to download: %s.%s%n", i.incrementAndGet(), v, t.getMessage(), threadName);
+                return null;
+            } finally {
+                lock.unlock();
+            }
```

Comment:
callout: I'd almost expect for this to be written out by the main thread once all of the downloads were completed. Given the use of a concurrent map, that would also alleviate as much contention on the lock which isn't needed -- especially with just the download, then the write of the CSV by the main thread.